### PR TITLE
21154 bulk actions timing out

### LINKED ISSUE #21154: [Bulk Actions] : Executing action on many contents can get timed out in cloud instances
Found when working on Support: [105339](https://dotcms.zendesk.com/agent/tickets/105339)
Reproducible as of dotCMS 21.06.3, up to latest master (21.10)

There are several customers that have several thousands of contents of a given type. If you try to execute a given Workflow Action on all of them, the bulk action dialog will time out and the summary table with the results of the execution will not be displayed. **This problem will NOT happen locally**. For security reasons, the default timeouts for development environments are not as strict as the ones used in Cloud instances, so the official fix must be tested in there.

**To Reproduce**
Reach out to me to get more information on a data set that has large amounts of contents for a given type.
1. Go to the Content Search.
2. Select a Content Type that has, ideally, thousands of contents.
3. Select all published contents from it.
4. Create a custom Workflow Action that un-publishes/archives/deletes them. This is important so that we can make sure that dotCMS will take longer to execute all actions on the contents.
5. Run the action, and wait for about 1 minute. You'll see the `Unable to show available actions` in the section where the summary table should be displayed:

img-5

**Expected behavior**
The bulk actions dialog should not time out as it provides extremely important information on failed contents. It needs to wait for the action to finish so that Users can deal with the failed records.

**Additional context**
We have several customers with very big data repositories. Improving the way the bulk actions dialog works is very important.

### PR DESCRIPTION
**Pull request by dsilvam**
With this PR, Server-Sent Events are introduced via a new endpoint for executing bulk workflow actions. 
A caller can then subscribe to the event producer and receive events and act accordingly. 
To make possible the support of POST requests a library call sse.js was included.

Changes include:
* A new endpoint implementing Server-Sent Events, under `workflow/contentlet/actions/_bulkfire`
* New non-blocking methods at API level for firing bulk WF actions
* Changes in JSP to subscribe to the event producer and process events

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPI.java [hunk 1]
```diff
@@ -1079,7 +1085,14 @@ WorkflowTask createWorkflowTask(final Contentlet contentlet, final User user,
 	Map<String, List<WorkflowScheme>> findSchemesMapForContentType(List<ContentType> contentTypes)  throws DotDataException;
 
 
-
+	void fireBulkActionTasks(final WorkflowAction action,
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPI.java:1088]: Add doc

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/workflow/WorkflowResource.java [hunk 2]
```diff
@@ -564,6 +572,97 @@ public final void fireBulkActions(@Context final HttpServletRequest request,
...
+    @Path("/contentlet/actions/_bulkfire")
+    @JSONP
+    @Produces(SseFeature.SERVER_SENT_EVENTS)
+    public EventOutput fireBulkActions(@Context final HttpServletRequest request,
+            final FireBulkActionsForm fireBulkActionsForm) {
+        final InitDataObject initDataObject = this.webResource
+                .init(null, request, new EmptyHttpResponse(), true, null);
+
+        final EventOutput eventOutput = new EventOutput();
+        new Thread(new Runnable() {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/workflow/WorkflowResource.java:585]: use the DotConcurrent

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/workflow/WorkflowResource.java [hunk 3]
```diff
@@ -564,6 +572,97 @@ public final void fireBulkActions(@Context final HttpServletRequest request,
...
+    @POST
+    @Path("/contentlet/actions/_bulkfire")
+    @JSONP
+    @Produces(SseFeature.SERVER_SENT_EVENTS)
+    public EventOutput fireBulkActions(@Context final HttpServletRequest request,
+            final FireBulkActionsForm fireBulkActionsForm) {
+        final InitDataObject initDataObject = this.webResource
+                .init(null, request, new EmptyHttpResponse(), true, null);
+
+        final EventOutput eventOutput = new EventOutput();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/workflow/WorkflowResource.java:584]: I think this is ok but eventually you can encapsulate this on an API or Delegate such as
```
ServerSideEventDelegate delegate = SystemEventsAPI.createServerSideEventDelegate();
// .....
delegate.sendMap(map("success", delta));
// .....
delegate.sendString("failed:" + inode));
```

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java [hunk 4]
```diff
@@ -2827,6 +2979,136 @@ private BulkActionsResultView distributeWorkAndProcess(final WorkflowAction acti
...
+					future.get();
+				} catch (InterruptedException | ExecutionException e) {
+					Logger.error(this, e.getMessage(), e);
+				}
+			}
+
+			Try.run(() -> {
+				APILocator.getNotificationAPI().generateNotification(new I18NMessage("Workflow-Bulk-Actions-Title"), // title = Reindex Notification
+						new I18NMessage("Workflow-Bulk-Actions-Finished", null, null), null, // no actions
+						NotificationLevel.INFO, NotificationType.GENERIC, Visibility.ROLE, cmsAdminRole.getId(), systemUser.getUserId(),
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java:3106]: Send the notification to the user who ran the bulk action only

**Review by wezell — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/webapp/html/portlet/ext/contentlet/view_bulk_actions_inc.jsp [hunk 5]
```diff
@@ -332,23 +336,36 @@
...
+
+        var entity = {skippedCount: 0, successCount: 0, fails: 0};
+
+        bulkWorkflowActionCallback(entity);
+        successCount = 0;
+        failCount = 0;
+
+        source.addEventListener('success', function(e) {
+            // Assuming we receive JSON-encoded data payloads:
+            console.log(e.data);
```
Comment (wezell) [inline dotCMS/src/main/webapp/html/portlet/ext/contentlet/view_bulk_actions_inc.jsp:353]: remove `console.log` from the production code

**Review by wezell — CHANGES_REQUESTED** [review]

**Review by erickgonzalez — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java [hunk 6]
```diff
@@ -2674,6 +2758,33 @@ private BulkActionsResultView fireBulkActionsTaskForSelectedInodes(final Workflo
...
+			throws DotDataException {
+
+		try {
+
+			final String luceneQuery = String
+					.format("+inode:( %s ) +(wfstep:%s )", String.join(StringPool.SPACE, inodes),
+							String.join(" wfstep:", workflowAssociatedStepIds));
+
+			final String sanitizedQuery = LuceneQueryUtils.sanitizeBulkActionsQuery(luceneQuery);
+			//final long skipsCount = (inodes.size() - contentlets.size());
```
Comment (victoralfaro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java:2777]: Nitpick: remove unused

**Review by victoralfaro-dotcms — APPROVED** [review]
LGTM

**Pull request merged**